File: assembler folder/parser.py
class Parser:
    def __init__(self, file_path):
        self.file_path = file_path
        self.current_command = None
        self.current_line = 0
        with open(file_path, 'r') as file:
            self.commands = [line.strip() for line in file]
            self.commands = [cmd.split('//')[0].strip() for cmd in self.commands]
            self.commands = [cmd for cmd in self.commands if cmd]
    
    def has_more_commands(self):
        return self.current_line < len(self.commands)
    
    def advance(self):
        if self.has_more_commands():
            self.current_command = self.commands[self.current_line]
            self.current_line += 1
    
    def command_type(self):
        command = self.current_command.strip()
        if command.startswith('(') and command.endswith(')'):
            return 'L_COMMAND'
        elif command.startswith('@'):
            return 'A_COMMAND'
        else:
            return 'C_COMMAND'
    
    def symbol(self):
        if self.command_type() == 'A_COMMAND':
            return self.current_command[1:]
        elif self.command_type() == 'L_COMMAND':
            return self.current_command[1:-1]
        return None
    
    def dest(self):
        if '=' in self.current_command:
            return self.current_command.split('=')[0]
        return 'null'
    
    def comp(self):
        if '=' in self.current_command:
            command = self.current_command.split('=')[1]
        else:
            command = self.current_command
        if ';' in command:
            return command.split(';')[0]
        return command
    
    def jump(self):
        if ';' in self.current_command:
            return self.current_command.split(';')[1]
        return 'null'

File: assembler folder/Parser.java
import java.io.*;
import java.util.*;

public class Parser {
    private List<String> commands;
    private String currentCommand;
    private int currentLine;

    public Parser(String filePath) throws IOException {
        this.commands = new ArrayList<>();
        this.currentCommand = null;
        this.currentLine = 0;

        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                if (!line.isEmpty()) {
                    // Remove comments
                    String[] parts = line.split("//");
                    line = parts[0].trim();
                    if (!line.isEmpty()) {
                        commands.add(line);
                    }
                }
            }
        }
    }

    public boolean hasMoreCommands() {
        return currentLine < commands.size();
    }

    public void advance() {
        if (hasMoreCommands()) {
            currentCommand = commands.get(currentLine);
            currentLine++;
        }
    }

    public String commandType() {
        String command = currentCommand.trim();
        if (command.startsWith("(") && command.endsWith(")")) {
            return "L_COMMAND";
        } else if (command.startsWith("@")) {
            return "A_COMMAND";
        } else {
            return "C_COMMAND";
        }
    }

    public String symbol() {
        if (commandType().equals("A_COMMAND")) {
            return currentCommand.substring(1);
        } else if (commandType().equals("L_COMMAND")) {
            return currentCommand.substring(1, currentCommand.length() - 1);
        }
        return null;
    }

    public String dest() {
        if (currentCommand.contains("=")) {
            return currentCommand.split("=")[0];
        }
        return "null";
    }

    public String comp() {
        String command;
        if (currentCommand.contains("=")) {
            command = currentCommand.split("=")[1];
        } else {
            command = currentCommand;
        }
        if (command.contains(";")) {
            return command.split(";")[0];
        }
        return command;
    }

    public String jump() {
        if (currentCommand.contains(";")) {
            return currentCommand.split(";")[1];
        }
        return "null";
    }
}